File: leetcode/201-300/233-number-of-digit-one/bench/digit_one.c
#include <stdio.h>

static long count_digit_one(long n) {
    if (n < 0) return 0;
    long count = 0;
    long pos = 1;
    while (pos <= n) {
        long high = n / (pos * 10);
        long cur = (n / pos) % 10;
        long low = n % pos;
        if (cur == 0) {
            count += high * pos;
        } else if (cur == 1) {
            count += high * pos + low + 1;
        } else {
            count += (high + 1) * pos;
        }
        pos *= 10;
    }
    return count;
}

int main(void) {
    long limit = 6000000;
    long sink = 0;
    for (long i = 0; i < limit; i++) {
        sink += count_digit_one(i);
    }
    printf("%ld\n", sink);
    return 0;
}

File: leetcode/201-300/233-number-of-digit-one/bench/digit_one.py
"""Benchmark workload for LeetCode #233 — Number of Digit One (Python; scale lane)."""


def count_digit_one(n):
    if n < 0:
        return 0
    count = 0
    pos = 1
    while pos <= n:
        high = n // (pos * 10)
        cur = (n // pos) % 10
        low = n % pos
        if cur == 0:
            count += high * pos
        elif cur == 1:
            count += high * pos + low + 1
        else:
            count += (high + 1) * pos
        pos *= 10
    return count


def main():
    limit = 6000000
    sink = 0
    for i in range(limit):
        sink += count_digit_one(i)
    print(sink)


main()
